neuralmonkey/encoders/imagenet_encoder.py

```diff
@@ -97,56 +98,70 @@ def __init__(self,
-                        ("Endpoint '{}' for network '{}' cannot be "
-                         "a convolutional map, its dimensionality is: {}."
-                        ).format(attention_layer, network_type,
-                                 ", ".join([str(d.value) for d in
-                                            net_output.get_shape()])))
-
-                if not fine_tune:
-                    net_output = tf.stop_gradient(net_output)
-                # pylint: disable=no-member
-                shape = [s.value for s in net_output.get_shape()[1:]]
-                # pylint: enable=no-member
-                self.__attention_tensor = tf.reshape(
-                    net_output, [-1, shape[0] * shape[1], shape[2]])
-
-            if encoded_layer is not None:
-                if encoded_layer not in end_points:
-                    raise ValueError(
-                        "Network '{}' does not contain endpoint '{}'.".format(
-                            network_type, encoded_layer))
-
-                self.encoded = tf.squeeze(end_points[encoded_layer], [1, 2])
-                if not fine_tune:
-                    self.encoded = tf.stop_gradient(self.encoded)
-            else:
-                self.encoded = tf.reduce_mean(net_output, [1, 2])
-    # pylint: enable=too-many-arguments,too-many-locals
+            _, self._end_points = net_function(self.input_image)
+
+    @tensor
+    def input_image(self) -> tf.Tensor:
+        return tf.placeholder(
+            tf.float32, [None, self.HEIGHT, self.WIDTH, 3])
+
+    @tensor
+    def cnn_states(self) -> tf.Tensor:
+        if self._attention_layer is None:
+            return None
+
+        if self._attention_layer not in self._end_points:
+            raise ValueError(
+                "Network '{}' does not contain endpoint '{}'".format(
+                    self._network_type, self._attention_layer))
+
+        net_output = self._end_points[self._attention_layer]
+
+        if len(net_output.get_shape()) != 4:
+            raise ValueError(
+                ("Endpoint '{}' for network '{}' cannot be "
+                 "a convolutional map, its dimensionality is: {}."
+                ).format(self._attention_layer, self._network_type,
+                         ", ".join([str(d.value) for d in
+                                    net_output.get_shape()])))
+
+        if not self._fine_tune:
+            net_output = tf.stop_gradient(net_output)
+        return net_output
+
+    @tensor
+    def states(self) -> tf.Tensor:
+        # pylint: disable=no-member
+        shape = [s.value for s in self.cnn_states.get_shape()[1:]]
+        # pylint: enable=no-member
+        return tf.reshape(
+            self.cnn_states, [-1, shape[0] * shape[1], shape[2]])
+
+    @tensor
+    def encoded(self) -> tf.Tensor:
+        if self._encoded_layer is None:
+            return tf.reduce_mean(self.cnn_states, [1, 2])
+
+        if self._encoded_layer not in self._end_points:
```

Comment:
zase - kde se vemou end points a jestli to neni tenzor tak bych to dal do konstruktoru